### PR TITLE
Fix `BatchSizeFinder` leaving model in train state

### LINKED ISSUE #18813: BatchSizeFinder leaves model in the train state if used with trainer.validate
### Bug description

If `BatchSizeFinder` is used with `trainer.validate` or `trainer.test` it would leave model in the train state, therefore enabling Dropout and other model randomness. That would influence model predictions and would result in unreliable validation metrics produced.

Step to reproduce:
1. Add `BatchSizeFinder` to a trainer 
2. Run validation (without fit)
3. See how model output changes from one without `BatchSizeFinder` (can also be influenced by changing random state)

While this doesn't matter for `trainer.fit` (or LightningCLI `fit` arg for that matter) it will create undesired randomness when user want to reevaluate a model they already trained
 

### What version are you seeing the problem on?

v1.8, v2.0

### How to reproduce the bug

Minimal example with 3 validation outputs:
1. Without `BatchSizeFinder`
2. With default `BatchSizeFinder`
3. With `BatchSizeFinder` that calls to `trainer.model.eval()`

Note that something like `Dropout` need to be present in the model to replicate this behavior.

```python
import os

import torch
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import BatchSizeFinder
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 128)
        self.dropout = torch.nn.Dropout(0.5)
        self.batch_size = 1

    def forward(self, x):
        return self.dropout(self.layer(x))

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


class CustomBatchSizeFinder(BatchSizeFinder):
    def on_validation_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        super().on_validation_start(trainer, pl_module)
        trainer.model.eval()


def run():
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    print("Without BatchSizeFinder:")
    trainer.validate(model, dataloaders=val_data)

    callbacks = [BatchSizeFinder()]
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        callbacks=callbacks
    )
    print("With BatchSizeFinder:")
    trainer.validate(model, dataloaders=val_data)

    callbacks = [CustomBatchSizeFinder()]
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        callbacks=callbacks
    )
    print("With BatchSizeFinder that calls to `trainer.model.eval()`:")
    trainer.validate(model, dataloaders=val_data)


if __name__ == "__main__":
    torch.manual_seed(1)
    run()
```

### Error messages and logs

```
Without BatchSizeFinder:
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     Validate metric           DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       valid_loss           15.337783813476562
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
With BatchSizeFinder:
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     Validate metric           DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       valid_loss           26.495624542236328
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
With BatchSizeFinder that calls to `trainer.model.eval()`:
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     Validate metric           DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       valid_loss           15.337783813476562
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
As you can see the loss with `BatchSizeFinder` differs from other two options
Difference in loss can be fixed either by calling `trainer.model.eval()` OR by removing randomness from the model (Dropout in that case)

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 3050 Laptop GPU
        - available:         True
        - version:           12.1
* Lightning:
        - lightning:         2.1.0
        - lightning-utilities: 0.9.0
        - pytorch-lightning: 2.1.0
        - torch:             2.1.0
        - torchmetrics:      1.2.0
* Packages:
        - aiohttp:           3.8.6
        - aiosignal:         1.3.1
        - async-timeout:     4.0.3
        - attrs:             23.1.0
        - certifi:           2023.7.22
        - charset-normalizer: 3.3.0
        - filelock:          3.12.4
        - frozenlist:        1.4.0
        - fsspec:            2023.9.2
        - idna:              3.4
        - jinja2:            3.1.2
        - lightning:         2.1.0
        - lightning-utilities: 0.9.0
        - markupsafe:        2.1.3
        - mpmath:            1.3.0
        - multidict:         6.0.4
        - networkx:          3.1
        - numpy:             1.24.4
        - nvidia-cublas-cu12: 12.1.3.1
        - nvidia-cuda-cupti-cu12: 12.1.105
        - nvidia-cuda-nvrtc-cu12: 12.1.105
        - nvidia-cuda-runtime-cu12: 12.1.105
        - nvidia-cudnn-cu12: 8.9.2.26
        - nvidia-cufft-cu12: 11.0.2.54
        - nvidia-curand-cu12: 10.3.2.106
        - nvidia-cusolver-cu12: 11.4.5.107
        - nvidia-cusparse-cu12: 12.1.0.106
        - nvidia-nccl-cu12:  2.18.1
        - nvidia-nvjitlink-cu12: 12.2.140
        - nvidia-nvtx-cu12:  12.1.105
        - packaging:         23.2
        - pip:               23.2.1
        - pytorch-lightning: 2.1.0
        - pyyaml:            6.0.1
        - requests:          2.31.0
        - setuptools:        68.1.2
        - sympy:             1.12
        - torch:             2.1.0
        - torchmetrics:      1.2.0
        - tqdm:              4.66.1
        - triton:            2.1.0
        - typing-extensions: 4.8.0
        - urllib3:           2.0.7
        - wheel:             0.41.2
        - yarl:              1.9.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.18
        - release:           5.15.0-83-generic
        - version:           #92-Ubuntu SMP Mon Aug 14 09:30:42 UTC 2023

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by tanaymeh**
## What does this PR do?
This PR patches the bug where `BatchSizeFinder` would leave the model in train state if used with trainer.validate

Fixes #18813

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review
@BoringDonut

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18826.org.readthedocs.build/en/18826/

### CODE REVIEW

**Review by BoringDonut — CHANGES_REQUESTED** [review]
@tanaymeh probably missed new `is_training` arg in `on_fit_start`
```
  File "lightning/pytorch/callbacks/batch_size_finder.py", line 187, in on_fit_start
    self.scale_batch_size(trainer, pl_module)
TypeError: scale_batch_size() missing 1 required positional argument: 'is_training'
```

**Comment by BoringDonut** [comment]
Other then aforementioned missing arg, this PR seems to fix the issue for val/test/prediction, thanks a lot

**Comment by tanaymeh** [comment]
@BoringDonut Thanks for the review! I have added the requested changes, can you please review it again?

Thanks again!

**Review by BoringDonut — APPROVED** [review]
Seems to work well and fixed the issue. Tested with `trainer.fit` `.predict` `.validate` and `.test`
Presumably this PR doesn't change any behavior withing `BatchSizeFinder` internals, since it only affects the model after `_scale_batch_size` already did its main task anyway

File: src/lightning/pytorch/callbacks/batch_size_finder.py [hunk 1]
```diff
@@ -168,31 +168,36 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: O
...
+    def scale_batch_size(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", is_training: bool) -> None:
         new_size = _scale_batch_size(
             trainer,
             self._mode,
             self._steps_per_trial,
             self._init_val,
             self._max_trials,
             self._batch_arg_name,
+            is_training,
         )
```
Comment (BoringDonut) [inline src/lightning/pytorch/callbacks/batch_size_finder.py:180]: @tanaymeh On a second thought, I think we can avoid adding a new argument here

`pl_module` passed to `scale_batch_size` anyway, so there are no need to read `pl_module.training` earlier

Example:
```python3
    def scale_batch_size(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        is_training = pl_module.training
        new_size = _scale_batch_size(
            trainer,
            self._mode,
            self._steps_per_trial,
            self._init_val,
            self._max_trials,
            self._batch_arg_name,
            is_training,
        )
...
    def on_fit_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        self.scale_batch_size(trainer, pl_module)
...
```

Doing it that way would keep compatibility for people who might have called to `scale_batch_size` in their own code (e.g. classes inherited from `BatchSizeFinder`)

**Comment by awaelchli** [comment]
@tanaymeh @BoringDonut have you found where the batch size tuner calls `model.train()`? Ideally, we would want to make the finder not change it in the first place. This would be the real fix IMO. We should leave the responsibility of calling model.train()/eval() to the respective loop IMO.

**Comment by tanaymeh** [comment]
> @tanaymeh @BoringDonut have you found where the batch size tuner calls `model.train()`? Ideally, we would want to make the finder not change it in the first place. This would be the real fix IMO. We should leave the responsibility of calling model.train()/eval() to the respective loop IMO.

@awaelchli I didn't find out the place where the bug is happening but I am working on it!

**Comment by BoringDonut** [comment]
> have you found where the batch size tuner calls `model.train()`?

@tanaymeh @awaelchli 

I seems to found it
Call trace looks like that

1.https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/tuner/batch_size_scaling.py#L333
which redirects to 
2. https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/loops/evaluation_loop.py#L108
Last line of which calls to `return self.on_run_end()`
3. https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/loops/evaluation_loop.py#L246
Aaaaand this methods assumes it runs within the train loop, so it converts model to the train state
4.
https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/loops/evaluation_loop.py#L270
This `_on_evaluation_model_train` ends up in 
5. https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/core/hooks.py#L166



Basically it seems `BatchSizeFinder` should either not call `_EvaluationLoop` through `.run()`, as this method would unconditionally turn model to a train state
Or it indeed should compensate by calling to `model.eval` if it understands that it works with an `_EvaluationLoop`
I don't see an easy way to modify `_EvaluationLoop.run()` to address it without too much of a hustle 
As such, i think solution by @tanaymeh looks reasonable.
I guess to be more consistent with existing `BatchSizeFinder` style this call to `.eval` can be done in [BatchSizeFinder.__scale_batch_restore_params](https://github.com/Lightning-AI/lightning/blob/e7afe04ee86b64c76a5446088b3b75d9c275e5bf/src/lightning/pytorch/tuner/batch_size_scaling.py#L141)

**Comment by tanaymeh** [comment]
@BoringDonut I added your previously suggested changes of moving the `is_training` inside the `scale_batch_size` function.

Please let me know if I should add any more changes

cc: @awaelchli

File: src/lightning/pytorch/callbacks/batch_size_finder.py [hunk 2]
```diff
@@ -169,13 +169,15 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: O
             )
 
     def scale_batch_size(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
+        is_training = pl_module.training
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/batch_size_finder.py:172]: how about we remove the changes in this file ...

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 3]
```diff
@@ -32,6 +32,7 @@ def _scale_batch_size(
     init_val: int = 2,
     max_trials: int = 25,
     batch_arg_name: str = "batch_size",
+    is_training: bool = True,
```
Comment (awaelchli) [inline src/lightning/pytorch/tuner/batch_size_scaling.py:35]: ... and this argument too, but instead capture the `pl_module.training` state at the beginning of this function ...

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 4]
```diff
@@ -95,6 +96,12 @@ def _scale_batch_size(
     trainer._checkpoint_connector.restore(ckpt_path)
     trainer.strategy.remove_checkpoint(ckpt_path)
 
+    # Set the model to training or evaluation mode based on the is_training parameter
```
Comment (awaelchli) [inline src/lightning/pytorch/tuner/batch_size_scaling.py:99]: ... and restore it here

File: src/lightning/pytorch/tuner/batch_size_scaling.py (hunk 3 above)
Comment (tanaymeh) [reply src/lightning/pytorch/tuner/batch_size_scaling.py:35]: Thanks for the review! `pl_module` is never passed onto the `_scale_batch_size` function but I have accessed the training state from the `trainer`.

**Review by BoringDonut — APPROVED** [review]

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 5]
```diff
@@ -63,6 +63,9 @@ def _scale_batch_size(
         rank_zero_warn("Skipping batch size scaler since `fast_dev_run` is enabled.")
         return None
 
+    # Capture whether the module is in training or eval mode
+    is_training = trainer.training
```
Comment (carmocca) [inline src/lightning/pytorch/tuner/batch_size_scaling.py:67]: I would prefer that this captures the `trainer.lightning_module.training` state instead, which can still be used to restore later with `trainer.lightning_module.train(was_training)`

**Review by carmocca — COMMENTED** [review]
Do you think you can add a test to `tests/tests_pytorch/tuner/test_scale_batch_size.py`?

**Comment by tanaymeh** [comment]
> Do you think you can add a test to `tests/tests_pytorch/tuner/test_scale_batch_size.py`?

@carmocca Should the test be similar to @BoringDonut's script ([here](https://github.com/Lightning-AI/lightning/issues/18813#issue-1948086458)) that is used to test if the model produces the same output in train and eval modes?

**Comment by carmocca** [comment]
I would suggest calling `_scale_batch_size` directly with some unittest Mocks, and assert that the original training state is preserved

**Comment by tanaymeh** [comment]
> I would suggest calling `_scale_batch_size` directly with some unittest Mocks, and assert that the original training state is preserved

Should create a mock for the trainer and lightning module as well?

```python
from unittest.mock import Mock

...

trainer = Mock()
lightning_module = Mock()
trainer.lightning_module = lightning_module

...
```
